Stack switching proposal support

dhil opened the pull request:
This patch implements text and binary encoding/decoding support for the stack switching proposal. It does so by adapting the previous typed-continunations implementation. Particular changes:

* Support for new `resume` encoding.
* Added support for `resume_throw` and `switch`.
* Feature flag `typed-continuations` has been renamed to `stack-switching`.

A small unfortunate implementation detail is that the internal name `Switch` was already taken by the `br_table` instruction, so I opted to give the `switch` instruction the internal name `StackSwitch`.

A minor detail is that I have reordered the declarations/definitions of the stack switching instructions such that they appear in ascending order according to their opcode value (this is the same order that the stack-switching explainer document present them in).

I can look into adding validation support in a subsequent patch.

tlively commented on src/parser/parsers.h:
```diff
@@ -2432,42 +2445,85 @@ makeContBind(Ctx& ctx, Index pos, const std::vector<Annotation>& annotations) {
-    auto label = labelidx(ctx);
-    CHECK_ERR(label);
-    ctx.appendTagLabel(tagLabels, *tag, *label);
+    auto keyword = ctx.in.peekKeyword();
+    if (keyword == "switch") {
+      ctx.in.takeKeyword();
```
```suggestion
    if (ctx.in.takeKeyword("switch")) {
```

tlively commented on src/parser/parsers.h:
```diff
@@ -2432,42 +2445,85 @@ makeContBind(Ctx& ctx, Index pos, const std::vector<Annotation>& annotations) {
+  CHECK_ERR(exnTag);
+
+  auto resumetable = ctx.makeOnClauseList();
+  while (ctx.in.takeSExprStart("on"sv)) {
```
It would be good to deduplicate this code into a separate parser function for resume handler lists.

tlively commented on src/wasm/wasm-ir-builder.cpp:
```diff
@@ -1924,49 +1936,101 @@ Result<> IRBuilder::makeContBind(HeapType contTypeBefore,
   Resume curr(wasm.allocator);
   curr.contType = ct;
   curr.operands.resize(ct.getContinuation().type.getSignature().params.size());
+  curr.handlerTags.set(tags);
+  curr.onTags.set(onTags);
```
These don't affect how children get popped off the stack, so there's no need to set them here. In the loop below we can use `onTags` instead of `curr.onTags`.

tlively commented on src/wasm/wasm-ir-builder.cpp:
```diff
@@ -1924,49 +1936,101 @@ Result<> IRBuilder::makeContBind(HeapType contTypeBefore,
+    return Err{"arity mismatch: the continuation argument must have, at least, "
+               "unary arity"};
+  }
+  curr.operands.resize(nparams - 1); // the continuation argument of
+                                     // the continuation is synthetic,
+                                     // i.e. it is provided by the
+                                     // runtime.
```
Please put this comment on its own lines.

tlively commented on src/wasm/wasm-validator.cpp:
```diff
@@ -3495,43 +3508,58 @@ void FunctionValidator::visitContBind(ContBind* curr) {
+    !getModule() || getModule()->features.hasStackSwitching(),
     curr,
-    "resume requires typed-continuatons [--enable-typed-continuations]");
+    "resume_throw requires stack-switching [--enable-stack-switching]");
```
Let's check that exceptions are enabled as well.

dhil replied:
Good point! Thanks.

tlively reviewed (commented):
Thanks for working on this! Happy to meet and discuss any of these comments in real time if it would be more efficient for you.

tlively commented:
Also, if you could avoid force pushes as you update the PR, that will help me see what changes from version to version.

dhil commented:
> Also, if you could avoid force pushes as you update the PR, that will help me see what changes from version to version.

Yes, I only did that while rebasing and noone were reviewing :-) I will implement your feedback and push it directly on top of here.

dhil commented on src/wasm/wasm-binary.cpp:
```diff
@@ -7962,17 +7999,79 @@ void WasmBinaryReader::visitResume(Resume* curr) {
 }
 
-void WasmBinaryReader::visitSuspend(Suspend* curr) {
+void WasmBinaryReader::visitResumeThrow(ResumeThrow* curr) {
```
I think most (maybe all?) the shared logic concerns resume tables (i.e. `on` clauses). How about I abstract `handlerTags`, `onTags`, and `labels` into a new structure `ResumeTable`, which we attach to both `Resume` and `ResumeThrow`?

tlively replied:
The only problem with using a new kind of structure is that wasm-delegations-fields.def can really only handle Expression fields with very basic types. Because of that, it's probably better to keep the field definitions duplicated and just unify the logic that operates on them. We do this in a few places for the different call instructions, for instance.

dhil replied:
OK. If I template a, say, `ResumeExpression` class and let `Resume` and `ResumeThrow` subclass from it, then I can get wasm-delegations-fields.def to correctly map over the inherited (common) fields, right?

tlively replied:
I'm thinking more along the lines of having `visitResume` and `visitResumeThrow` call out to `template<typename T> void handleResumeClauses(T* curr) { ... }`.

dhil replied:
I've implemented your suggestion. I think we could push every into a templated function, if we allow ourselves to use some `constexpr` to check for `ResumeThrow` to handle extra field it has.

tlively replied:
I'm happy with how it is now. Better to abstract the right amount rather than over-extract and have to punch holes in the abstraction.

dhil commented on src/wasm.h:
```diff
@@ -1935,26 +1948,31 @@ class ContBind : public SpecificExpression<Expression::ContBindId> {
+  // Empty name (i.e. `Name()`) for switch tags.
   ArenaVector<Name> handlerBlocks;
+  // False if (on $tag $label) and true when (on $tag switch).
+  ArenaVector<bool> onTags;
```
I prefer the enum-approach, I think. It is close to what I did previously in wasm-tools. 

I modelled the current approach on how `TryTable` manages `catch_ref` and `catch_all_ref`.

dhil replied:
Commit https://github.com/WebAssembly/binaryen/pull/7041/commits/08864517e665ae9ec4ceeb7ef79b0a6b9dd9d4c3 uses the null-name trick to encode on-clauses.

dhil commented on src/wasm.h:
```diff
@@ -1923,6 +1925,17 @@ class StringSliceWTF : public SpecificExpression<Expression::StringSliceWTFId> {
+  ContNew() = default;
+  ContNew(MixedArena& allocator) {}
+
+  HeapType contType;
```
Yes, this is a good idea. I will look into.

dhil commented on src/wasm.h:
```diff
@@ -1972,16 +1990,47 @@ class Resume : public SpecificExpression<Expression::ResumeId> {
+public:
+  StackSwitch(MixedArena& allocator) : operands(allocator) {}
+
+  HeapType contType;
```
Here `contType` is not the type of `StackSwitch`, but rather the type of its input continuation. My current understanding of the inherited `type` member is that its role is to describe the type of the instruction.

tlively replied:
Ah, right. To clarify, we would be using `cont->type` as the source of truth and specially handling the case where `cont->type` is a bottom type or unreachable. This is what we do for e.g. `StructGet`, `ArrayGet`, `CallRef`, etc.

dhil commented on src/wasm/wasm.cpp:
```diff
@@ -1437,10 +1444,62 @@ void Resume::finalize(Module* wasm) {
+    auto& tgps = tagSig.params;
+    auto& tgrs = tagSig.results;
+
+    HeapType ftPrime{Signature(tgrs, ctrs)};
+    HeapType ctPrime{Continuation(ftPrime)};
+    Type ctPrimeRef(ctPrime, Nullability::NonNullable);
```
To calculate the continuation type passed to a branch target, we need to know the type of the label at the branch target, the incoming tag type, and the return type of `resume`.

tlively replied:
Yep, that seems like too much to recalculate in `finalize`.

dhil replied:
Just so I understand: do you propose to remove all the logic about (re)computing `sentTypes`?

tlively commented on src/passes/Print.cpp:
```diff
@@ -2824,7 +2858,7 @@ void PrintSExpression::visitTryTable(TryTable* curr) {
 
 void PrintSExpression::visitResume(Resume* curr) {
-  controlFlowDepth++;
+  // controlFlowDepth++;
```
Did you delete their declarations in `PrintSExpression` as well?

dhil replied:
Ah that's probably it. Thanks. I will give it a try tomorrow! I should be able to implement the rest of your feedback then.

dhil commented:
@tlively I think I have addressed all of your feedback now. Please take a look when you got time. Thanks in advance!

tlively commented on src/passes/Print.cpp:
```diff
@@ -301,7 +301,7 @@ struct PrintSExpression : public UnifiedExpressionVisitor<PrintSExpression> {
   void visitLoop(Loop* curr);
   void visitTry(Try* curr);
   void visitTryTable(TryTable* curr);
-  void visitResume(Resume* curr);
```
For the expressions that now depend on the types of their children to recover the printed type annotations, we now need to handle the case where the children are unreachable or have bottom heap types. See what we do for `StructNew` and friends just below this. It would also be good to add tests that exercise these cases.

dhil replied:
Please take a look at commits https://github.com/WebAssembly/binaryen/pull/7041/commits/157552ad4bcd77d7718fdfa944992b3505c416c4 and https://github.com/WebAssembly/binaryen/pull/7041/commits/480843d3b572e04b9b528b232af7dfc10971510e.

tlively commented on src/wasm/wasm-validator.cpp:
```diff
@@ -3477,61 +3479,79 @@ void FunctionValidator::visitStringSliceWTF(StringSliceWTF* curr) {
 }
 
-void FunctionValidator::visitContNew(ContNew* curr) {
+void FunctionValidator::visitSuspend(Suspend* curr) {
```
It would still be good to validate that `sendTypes` is the right size. We'll just have to require the sent types to be calculated when the expression is first built.

tlively commented on src/wasm/wasm.cpp:
```diff
@@ -1395,6 +1436,49 @@ void Resume::finalize(Module* wasm) {
     type = contSig.results;
   }
+
+  populateResumeSentTypes(this, wasm);
```
Hahaha, sorry for the back and forth. I think we're talking past each other here. I still think it doesn't make sense to populate the sent types in `finalize`. Instead, we should for the sent types to be populated from the start by whoever creates a `Resume` or `ResumeThrow` expression. In other words, `sentTypes` should be added as a required parameter to `Builder::makeResume` and `Builder::makeResumeThrow`.

dhil replied:
Yes, this was a mistake. I forgot to remove this code again after I reverted the commit. Please see commit https://github.com/WebAssembly/binaryen/pull/7041/commits/c8d345a900151fffe69df37d99a27b7076de3c26.

dhil commented on src/passes/Print.cpp:
```diff
@@ -380,6 +380,39 @@ struct PrintSExpression : public UnifiedExpressionVisitor<PrintSExpression> {
+    }
+  }
+  void visitContBind(ContBind* curr) {
+    if (!maybePrintUnreachableOrNullReplacement(
+          curr, Type(curr->sourceType, Nullable)) ||
```
I don't understand the rationale behind axing `sourceType` and `contType`. They represent immediates on the respective instructions. The immediates may point to types that differ from the types of the operands (in these cases `curr->cont`).

tlively replied:
The idea is that by recomputing the type immediates when necessary rather than caching them, we always produce the most precise possible types. If an optimization refines the type of the input to one of these instructions, then refinalization will refine the output type automatically. That couldn't happen if refinalization had to respect the cached type immediates.

dhil commented on test/lit/basic/stack_switching_resume_throw.wast:
Yes, thanks. I have added some now (I actually had them locally; I just forgot to add them to revision control).

dhil commented:
Do you prefer I resolve the merge conflict via rebase+force-push or just as a merge commit?

tlively commented on src/wasm/wasm-binary.cpp:
```diff
@@ -7981,6 +7981,52 @@ static void readResumeTable(WasmBinaryReader* reader, ResumeType* curr) {
+      auto& tgps = tagSig.params;
+      auto& tgrs = tagSig.results;
+
+      HeapType ftPrime{Signature(tgrs, ctrs)};
+      HeapType ctPrime{Continuation(ftPrime)};
+      Type ctPrimeRef(ctPrime, Nullability::NonNullable);
```
Shouldn't this be looking up the types based on the labels now?

dhil replied:
Is your point that we can simply read the type of the continuation off the (corresponding branch target) label type? I think so yes, it must have been given statically in the source code. It is not entirely clear to me what the purpose of `sentTypes` is (I suppose it is some caching mechanism).

tlively commented on src/wasm/wasm-ir-builder.cpp:
```diff
@@ -1945,6 +1945,55 @@ Result<> IRBuilder::makeSuspend(Name tag) {
+      auto& tgps = tagSig.params;
+      auto& tgrs = tagSig.results;
+
+      HeapType ftPrime{Signature(tgrs, ctrs)};
+      HeapType ctPrime{Continuation(ftPrime)};
+      Type ctPrimeRef(ctPrime, Nullability::NonNullable);
```
Same here.

dhil commented:
@tlively I am experiencing some difficulties with getting rid of the type immediate members (`contType`). Whenever I swap `contType` for `curr->cont->type.getHeapType()` in `child-typer.h` then I observe crashes in the `ConstraintCollector`. Supposedly, `curr->cont->type` is not always a continuation type... so something is wrong. I need another set of eyes to have a look. When you got time, do you mind taking a look at the code?

dhil commented:
Thanks for the pointer. I have sorted it out now, I think. The latest commit ought to eliminate the type immediates from the stack switching instructions.

tlively commented on test/lit/basic/stack_switching.wast:
```diff
@@ -0,0 +1,967 @@
+    (suspend $e1)
+  )
+
+  (func (export "unhandled-2")
```
The output will be more readable if all these functions have `$names` as well.

tlively reviewed (commented):
Looks great! Just a few comments on the tests.

dhil commented:
Hi @tlively, apologies for the delay, I didn't have time to look into finishing this patch up until now. Hopefully my latest commit addresses your feedback.

tlively reviewed (approved):
Amazing, thanks for all your work on this, @dhil!
